packages/cloudrouter/internal/cli/git_source.go

```
@@ -0,0 +1,147 @@
+			return "", fmt.Errorf("git URL must not contain credentials")
+		}
+	}
+	if parsed.RawQuery != "" || parsed.Fragment != "" || parsed.Opaque != "" || parsed.Path == "" {
+		return "", fmt.Errorf("git URL must not contain credentials, query parameters, or fragments")
+	}
+	if strings.ContainsAny(parsed.Host, "\x00\r\n\t") || strings.ContainsAny(parsed.Path, "\x00\r\n") {
+		return "", fmt.Errorf("git URL contains invalid control characters")
+	}
+	for _, component := range strings.Split(parsed.Path, "/") {
+		if component == ".." {
+			return "", fmt.Errorf("git URL path cannot contain ..")
```

golangci-lint (2.13.2)

[error] ST1005: error strings should not end with punctuation or newlines

(staticcheck)